[IMP] base: Make it easier to extend Attachment

keshrath opened the pull request:
Description of the issue/feature this PR addresses:

Fixes https://github.com/odoo/odoo/issues/33456

Current behavior before PR:

At the moment it is not easy to add a new storage type without overwriting several methods. The same applies to the method for migrating attachments.

Desired behavior after PR is merged:

This patch has the goal to improve this. An example that uses these changes is this addition of the storage type Large Object.

Large Objec: https://github.com/muk-it/muk_base/blob/12.0/muk_attachment_lobject/models/ir_attachment.py

Since the methods: `_get_datas_inital_vals`, `_update_datas_vals`, `_get_datas_clean_vals`, `_clean_datas_after_write` are used it would even work if several storage extensions were installed in the system, because now every module only has to implement its own cleanup method.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

keshrath commented:
There is another problem that this pull request does not solve: https://github.com/odoo/odoo/issues/33456

mart-e commented on odoo/addons/base/models/ir_attachment.py:
```diff
@@ -50,22 +50,62 @@ def _storage(self):
+            'file': ('store_fname', '=', False), 
+        }
+        record_domain = [
+            '&', ('type', '=', 'binary'),
```
that does not seems right, you will get something like
```python
[ '&', ('type', '=', 'binary'), '&', ('store_fname', '=', False), '|', ('res_field', '=', False), ('res_field', '!=', False)]
```

keshrath replied:
Isn't that the same as this form `[ '&', '&', ('type', '=', 'binary'), ('store_fname', '=', False), '|', ('res_field', '=', False), ('res_field', '!=', False)]` only in a different notation

mart-e commented:
Is this ready? You are not only creating overridable methods but also changing the logic on some places which seems odd.
Please explain why you are making these changes and how you would achieve to actually add a new storage method.

keshrath commented:
@mart-e 

First commit:

The only place where the logic should change is in the method `force_storage`. Since this method did not have the desired effect in my eyes. The rule (https://github.com/odoo/odoo/blob/b1596d182601bf4a81af7bd82323f7e9059fb571/odoo/addons/base/models/ir_attachment.py#L428) causes that not all attachments are beeing migrated. The separation into two methods (`force_storage` and `migrate`) is basically only for a better overview. 

The context (https://github.com/odoo/odoo/blob/b1596d182601bf4a81af7bd82323f7e9059fb571/odoo/addons/base/models/ir_attachment.py#L108) is necessary because otherwise a lot of mimetypes will break. The value `datas` triggers that the mimetypes are evaluated again, although nothing has changed. But now only `datas` are given as value so the recalculation is wrong in many cases.

The splitting of `_inverse_datas` into several methods should not change the logic and actually only serves for better extensibility.

Before:
https://github.com/muk-it/muk_base/blob/11.0/muk_attachment_lobject/models/ir_attachment.py

Afterwards:
https://github.com/muk-it/muk_base/blob/12.0/muk_attachment_lobject/models/ir_attachment.py

This solves the following scenario:

The filestore must be checked each time and the file may have to be deleted.

if fname:
    self._file_delete(fname)

This is not a major problem yet, because an extension could do the same, but if this extension also has a rule like this and there is another extension, it would have to take it into account or know about one another, but in my eyes this is very hard to maintain.

Second commit:

This commit addresses the problem described here: https://github.com/odoo/odoo/issues/33456

I do not find this solution particularly safe from a programming point of view.

https://github.com/muk-it/muk_base/blob/12.0/muk_attachment_lobject/models/ir_attachment.py#L111

The super method is also called here, but only works in this one file and not in models inherited from it.

https://github.com/odoo/odoo/pull/33392/files#diff-65be9bcf9d09c93a04039c00e1861b9fR248

I hope it is now a little clearer what the backgrounds of this pull request are. Please feel free to contact me if you have any questions.

keshrath commented:
@mart-e Did you have the opportunity to read my explanation and is it now clearer what this PR is trying to achieve? I have adapted the PR to the current master branch.

mart-e commented on odoo/addons/base/models/ir_attachment.py:
```diff
@@ -50,22 +50,62 @@ def _storage(self):
+        storage = self._storage().upper()
+        for index, attach in enumerate(self):
+            _logger.info(_("Migrate Attachment %s of %s to %s") % (index + 1, record_count, storage))
+            attach.with_context(migration=True).write({'datas': attach.datas})
```
and why no giving the existing mimetype in the write values? This way we avoid to introduce a context key to skip verification.

keshrath commented:
Thanks for taking the time to review. For some reason the source repository of this PR is broken and I can't make any more changes to it. So I created a new PR and incorporated the feedback and fixed the merge conflict.

PR: https://github.com/odoo/odoo/pull/35614#pullrequestreview-273379666
